flux-mini: add bulk job submission capabilities

grondo opened the pull request:
In the context of testing and exercising Flux interfaces, I've found it handy to have a tool that can submit a bunch of jobs rapidly, but our standard tools are definitely lacking in this department. I have a `bulksubmit.py` script which works, but it requires a directory full of jobspec files, which is not convenient.

As an experiment I made a small change to `flux-mini.py` to allow asynchronous submission of jobspec, and using that interface, added an option that allows bulk submission of the same job  rapidly. I'm putting up this experimental PR to get feedback before I move on to any testing.

This PR enhances the job submission used in `flux-mini.py` to use `submit_async()`, which opens the door to submitting multiple jobs per invocation (without performing them back to back).

Then, a new `--cc=IDSET` option is introduced to `flux job submit` which submits a copy of the requested job for each `id` in `IDSET`. The only difference is that the environment variable `FLUX_JOB_CC=id` is set for jobs submitted in this way, which allows each job to determine its index of submission if necessary (e.g. to operate on a different input file). This is similar, though not quite as efficient as, the [job array](https://slurm.schedmd.com/job_array.html) feature in Slurm.

e.g.

```
$ flux mini submit --cc=1-10 sh -c 'echo $FLUX_JOB_CC'
ƒLgfPweP1
ƒLgfPweP2
ƒLgfPweP3
ƒLgfPweP4
ƒLgfRRdfM
ƒLgfRRdfN
ƒLgfRRdfP
ƒLgfRRdfQ
ƒLgfRRdfR
ƒLgfRRdfS
$ flux job attach ƒLgfRRdfM
5
```

After this, it was trivial to add a `flux mini bulksubmit` command which, instead of taking a COMMAND and ARGS as positional parameters, treats each positional argument as separate command, and creates a job with the same submission arguments for each, e.g.:

```
$ flux mini bulksubmit hostname hostname hostname
ƒKfWcDdUb
ƒKfWdhckw
ƒKfWdhckx
```

or perhaps more useful

```
$ flux mini bulksubmit ./python/t*.py
ƒL1TnX4BH
ƒL1TnX4BJ
ƒL1TnX4BK
ƒL1Tp13Td
ƒL1Tp13Te
ƒL1Tp13Tf
ƒL1Tp13Tg
ƒL1Tp13Th
ƒL1TqV2jy
ƒL1TqV2jz
ƒL1TqV2k1
ƒL1TqV2k2
ƒL1Try22K
ƒL1Try22L
ƒL1Try22M
```

`flux mini bulksubmit` also supports the `--cc` option, in which case every command is CC'd for each ID in IDSET. This allows us, for instance, to submit a mix of sleep jobs very quickly:

```
$ time flux mini bulksubmit --cc=1-100 'sleep 0' 'sleep 1' 'sleep 3' | wc -l
300

real	0m2.219s
user	0m1.847s
sys	0m0.062s
```

Everything here is kind of open for discussion. The `--cc` option was at first `--repeat`, then `--replicate`, but I like how short `--cc` is, and how the exported variable is also short, since it might be used in scripts.

Also, the addition of a different `FLUX_JOB_CC` to each jobspec unfortunately means every jobspec is different. Maybe eventually we can figure out a way to have these kinds of jobs "share" a jobspec, in the meantime we could leave that out, or have it opt-in only.

garlick commented:
I think the interface is pretty awesome actually.

grondo commented:
I just pushed an update to add some very basic tests for `--cc` and `bulksubmit`, and also a bit of description of these new interfaces in `flux-mini(1)`.

Since these are new interfaces, we could merge this in before the tag in order to get wider distribution and possible feedback on this experiment before the next release. O/w, I'm fine waiting until just after the tag.

BTW, here's one fun use case for `bulksubmit`
```console
grondo@asp:~/git/flux-core.git/t$ ../src/cmd/flux start -s 2 sh -c 'flux mini bulksubmit ./t*.t ./python/t*.py; while flux jobs --stats-only; do sleep 2; done'
ƒWxYVH9
ƒWxYVHA
ƒWxYVHB
ƒWz2UZV
ƒWz2UZW
ƒWz2UZX
ƒWz2UZY
ƒWz2UZZ
ƒWz2UZa
ƒX1WTqq
[snip]

8 running, 0 completed, 0 failed, 117 pending
8 running, 0 completed, 0 failed, 117 pending
8 running, 1 completed, 0 failed, 116 pending
8 running, 3 completed, 0 failed, 114 pending
8 running, 7 completed, 0 failed, 110 pending
8 running, 9 completed, 0 failed, 108 pending
8 running, 11 completed, 0 failed, 106 pending
8 running, 11 completed, 0 failed, 106 pending
8 running, 12 completed, 0 failed, 105 pending
8 running, 12 completed, 0 failed, 105 pending
8 running, 16 completed, 0 failed, 101 pending
8 running, 18 completed, 0 failed, 99 pending
8 running, 21 completed, 0 failed, 96 pending
8 running, 21 completed, 0 failed, 96 pending
8 running, 23 completed, 0 failed, 94 pending
8 running, 25 completed, 0 failed, 92 pending
8 running, 25 completed, 0 failed, 92 pending
8 running, 25 completed, 0 failed, 92 pending
8 running, 26 completed, 0 failed, 91 pending
...
```

garlick reviewed (approved):
This is really neat, and I think it is pretty low risk if we wanted to get it in for the tag and then ask for feedback from users.  I'll go ahead and approve, based on a fairly quick skim.

grondo commented:
Ah, one reason to hold off is the kludge in commit d358115. @SteVwonder is looking into the root cause and I'll open a proper issue when I get back to a keyboard

dongahn commented:
Great work @grondo.

In terms of feedback, I'm wondering how much additional performance benefits this can provide when dealing with sub-second tasks. I remember there were a couple of real use case where python invocation overheads were on the critical path to dealing with running micro tasks...

chu11 commented on src/cmd/flux-mini.py:
```diff
@@ -679,6 +755,24 @@ def main():
+    bulksubmit_parser_sub = subparsers.add_parser(
+        "bulksubmit",
+        parents=[bulksubmit.get_parser()],
+        help="equeue jobs in bulk",
```
typo - `enqueue`

grondo replied:
Thanks! Forced a push with a fix.

chu11 reviewed (commented):
noticed one typo

grondo commented:
Thanks to feedback from @SteVwonder, I changed `gc.collect()` to `gc.collect(0)` in this working branch. That still fixes the dangling reference which caused `flux job submit` to hang in the reactor, but the shallower gc greatly improves the job submission rate with `bulksubmit` and `--cc`:

```console
$ time flux mini submit --cc=1-300 sleep 0 | wc -l
300

real	0m0.752s
user	0m0.348s
sys	0m0.031s

```

(Compare to 2.1s for 300 jobs in example above)

grondo commented:
As discussed at :coffee:, I just pushed my experimental `--wait` and `--progress` options for `flux mini submit`. The `--wait` option will wait for all submitted jobs to complete by watching the main job eventlog, with `-v` the command will also print all submitted jobs' output, e.g. for a single job:
```
ƒ(s=4,d=0,builddir) grondo@asp:~/git/flux-core.git$ flux mini submit --wait -v hostname
ƒ23prmSXk3
ƒ23prmSXk3: 0: asp
```

However, this is most useful with the `--cc` option or the `bulksubmit` command, e.g.:

```console
$ flux mini submit --wait --cc=1-10 -v hostname
ƒ25bUoVAzs
ƒ25bUoVAzt
ƒ25bUoVAzu
ƒ25bUpyAHD
ƒ25bUpyAHE
ƒ25bUpyAHF
ƒ25bUpyAHG
ƒ25bUpyAHH
ƒ25bUrT9ZZ
ƒ25bUrT9Za
ƒ25bUoVAzt: 0: asp
ƒ25bUoVAzs: 0: asp
ƒ25bUoVAzu: 0: asp
ƒ25bUpyAHD: 0: asp
ƒ25bUpyAHE: 0: asp
ƒ25bUpyAHH: 0: asp
ƒ25bUpyAHF: 0: asp
ƒ25bUpyAHG: 0: asp
ƒ25bUrT9Za: 0: asp
ƒ25bUrT9ZZ: 0: asp
```

The `--progress` option shows progress of either job submission without `--wait` or job completion with `--wait`, e.g. (sorry, asciicast2gif didn't render the progressbar too well)

![asciicast](https://user-images.githubusercontent.com/741970/103604505-2bc8c080-4ec6-11eb-938d-2e78cb6b2a9d.gif)

SteVwonder commented:
While using bulksubmit, I got an exception:
```
flux-mini: ERROR: 'Namespace' object has no attribute 'verbose'
flux-mini: DEBUG: Traceback (most recent call last):
  File "/usr/src/src/bindings/python/flux/util.py", line 166, in __call__
    main_func()
  File "/usr/src/src/cmd/flux-mini.py", line 902, in main
    args.func(args)
  File "/usr/src/src/cmd/flux-mini.py", line 582, in main
    self.run_and_exit()
  File "/usr/src/src/cmd/flux-mini.py", line 499, in run_and_exit
    self.flux_handle.reactor_run()
  File "/usr/src/src/bindings/python/flux/core/handle.py", line 288, in reactor_run
    Flux.raise_if_exception()
  File "/usr/src/src/bindings/python/flux/core/handle.py", line 122, in raise_if_exception
    raise cls.set_exception(None) from None
  File "/usr/src/src/bindings/python/flux/future.py", line 34, in continuation_callback
    py_future.then_cb(py_future, *py_future.then_args, **py_future.then_kwargs)
  File "/usr/src/src/cmd/flux-mini.py", line 402, in event_watch_cb
    if args.verbose:
AttributeError: 'Namespace' object has no attribute 'verbose'
```

The bulk submit class parser just needs a `verbose` option added.

grondo commented:
Oops, sorry! I also found the progress bar wasn't functional with `--cc`, and have fixed these issues in my working copy.

However, my working copy has drifted far from this experimental PR. Let me force push my latest (which still isn't fully functional, unfortunately).

grondo commented:
Ok, I've updated this WIP PR with the latest changes, though you are warned - this is still experimental.

This PR now introduces a new `flux.progress` module which contains two classes, a `Bottombar` class which uses vt100 escape codes to keep a statusbar on the last line of a tty by reducing the scroll region by 1 (heavily borrowed from [bottombar.py](https://github.com/evalf/bottombar)). A `ProgressBar` class is derived from `Bottombar` so that we have access to a progress bar which allows the terminal to continue scrolling (example as seen above). (There are plenty of other Python libs out there that can do this. However, it seemed important to have a simple version available for flux-core which didn't require pip install of dependencies)

The latest code also changes the way `bulksubmit` works. `flux mini bulksubmit` now works more like `xargs` or GNU parallel, in that it takes arguments from either stdin or the commandline, and substitutes each into the "initial arguments" provided on the command line.

Since we're dealing with Python, the arguments are substituted with `.format()`, which luckily allows `{}` to work as expected:

```
$ echo foo bar baz | flux mini bulksubmit --dry-run -vv echo {}
Submitting "echo foo"
Submitting "echo bar"
Submitting "echo baz"
```

`bulksubmit` also allows `args` to be supplied on the command line, borrowing syntax from GNU parallel to separate command and args with the string `:::`:

```
$ flux mini bulksubmit --dry-run -vv echo {} ::: foo bar baz
Submitting "echo foo"
Submitting "echo bar"
Submitting "echo baz"
```

Multiple sets of arguments are allowed, in which case `bulksubmit` generates all combinations. The current arg from each group is passed as positional parameters to the internal `.format()`, so `{0}` references the first arg `{1}` the second, etc:

```
$ flux mini bulksubmit --dry-run -vv echo {0} {1} ::: foo bar baz ::: 1 2 3
Submitting "echo foo 1"
Submitting "echo foo 2"
Submitting "echo foo 3"
Submitting "echo bar 1"
Submitting "echo bar 2"
Submitting "echo bar 3"
Submitting "echo baz 1"
Submitting "echo baz 2"
Submitting "echo baz 3"
```

I'm thinking of renaming `bulksubmit` to `xsub` as a nod to the idea it is inspired from xargs...

I've also introduced a `--bcc` option which is exactly the same as `--cc`, except that `FLUX_JOB_CC` is not set in each submitted jobspec, and therefore it is literally submitting copies of the same jobspec (likely only useful in testing).

The progress bar issue with `--cc` and `bulksubmit` should be fixed now, but there are now a lot of argument combinations to test, and I have not tested them all yet.

grondo commented:
Also, for testing purposes it was inconvenient to have output labeled by default with `--wait -v`. Therefore, by default output with `submit` and `bulksubmit`  `--wait -v` is only labeled with `--label-io`:

grondo commented:
Just pushed an enhancement based on a suggestion from @SteVwonder. `flux mini bulksubmit` args can also be substituted with a replacement format string, e.g.:

```console
$ flux mini bulksubmit --dry-run -vv -n {} hostname ::: {1..8}
Submitting "-n1 hostname"
Submitting "-n2 hostname"
Submitting "-n3 hostname"
Submitting "-n4 hostname"
Submitting "-n5 hostname"
Submitting "-n6 hostname"
Submitting "-n7 hostname"
Submitting "-n8 hostname"
```

All the following options support substitution currently (easily expanded):
```python
        mutable_args = [
            "ntasks",
            "nodes",
            "setopt",
            "setattr",
            "job_name",
            "input",
            "output",
            "error",
        ]
```

grondo commented:
Ok, I cleaned up this branch a little, added tests, and updated the manpage.

There's probably a lot more that could be done with `bulksubmit`, but I think this is enough for now, and the commands are very useful for testing, so it would be nice to have the basic support in upstream flux-core.

I attempted to rework my commits to avoid the one giant "add bulk job submission capabilities.." commit, but that was becoming a struggle. Hopefully the result is not too difficult to understand.

In the final version, several bulk submission related options are added to `flux mini submit`:

 * `--cc=IDSET`: Submit a job for each id in IDSET, with `FLUX_JOB_CC=id` set
 *  `--bcc=IDSET`: Like above, but not `FLUX_JOB_CC` set in environment, jobs are duplicates
 * `--wait`: wait for all submitted jobs to complete and exit with the highest job exit code (very useful in testing)
 * `--watch`: print all job stdout/err, implies `--wait`
 * `--progress`: Without `--wait` print a progress bar showing progress of job submission, with `--wait` progress is for job completion and additionally includes the count of jobs pending, running, completed, and failed.
 * `--jps`: with `--progress`, display a job throughput statistic with the progress bar

The newly added command `bulksubmit` works as reported above (as inspired by GNU parallel and xargs).
Most submission options are now "mutable" (i.e. can be substituted using replacement strings), there are some new default attributes available with the substitutions (e.g. `{.%}` returns the input with any filename extension removed, `{./}` returns the basename, etc.), and a `--define` option can be used to define new methods/substitutions available as `{.<name>}`.

e.g. for @SteVwonder's scaling study, you could do:
```
$ flux mini bulksubmit --dry-run --define=p2='2**int(x)' -n {.p2} myapp ::: $(seq 0 8)
flux-mini: submit -n1 myapp
flux-mini: submit -n2 myapp
flux-mini: submit -n4 myapp
flux-mini: submit -n8 myapp
flux-mini: submit -n16 myapp
flux-mini: submit -n32 myapp
flux-mini: submit -n64 myapp
flux-mini: submit -n128 myapp
flux-mini: submit -n256 myapp
```

and the whole reason this work started, running the flux testsuite as flux jobs while watching the output of every job and including progress can be done as:
```
ƒ(s=4,d=0,builddir) grondo@asp:~/git/flux-core.git/t$ flux mini bulksubmit --watch --progress ./{} ::: t[0-9]*.t python/t*.py lua/t*.t
[snip]

ok 23 - kvs-copy to namespace does not unlink src
ok 52 - kvs: unlink -R works
ok 2 - Hydra can launch Flux
# passed all 2 test(s)
1..2
ok 3 - module watcher gets disconnected on module unload
# passed all 3 test(s)
ok 8 - flux kvs eventlog wait-event detects eventlog with event
ok 25 - kvs: multi blob-ref valref with an empty blobref on right, can be read
ok 31 - kvs: setroot pause / unpause works (owner)
ok 32 - kvs: setroot pause / unpause works (user)
ok 55 - kvs: unlink works on link that points to invalid namespace
ok 30 - kvs: lookup - namespace specified in environment variable works
ok 31 - kvs: lookup - namespace specified in command line overrides environment variable
ok 48 - error: resource_exclusive_invalid.yaml
flux-start: Error: --caliper-profile used but no Caliper support found
1..0 # SKIP skipping Caliper tests, Flux not built with Caliper support
ok 9 - flux kvs eventlog wait-event outputs more events with -v
ok 33 - kvs-copy nonexistent src namespace fails

PD:88  R:16  CD:31  F:0   │████████▊                             │ 23.0% 0:00:57
```

I believe this is ready for an actual review.

garlick commented:
I guess I shouldn't be too surprised, but this runs things really, really fast!  I'm just getting used to the powerful command line syntax but I think like you said, this is a going to be a go to tool for on the fly soak/scalability testing.

garlick reviewed (approved):
I've played with this a bit and spun through the diffs and didn't see anything to comment on. 

My vote is merge this and sort out the pieces later!  It's an awesome tool.  Looks like some nice python building blocks were created along the way.

grondo commented:
> I'm just getting used to the powerful command line syntax but I think like you said, this is a going to be a go to tool for on the fly soak/scalability testing.

The use of GNU parallel syntax is no accident, btw. There are many tutorials about using `parallel` within Slurm jobs for high throughput computing, e.g.

https://curc.readthedocs.io/en/latest/software/GNUParallel.html#gnu-parallel
https://www.marcc.jhu.edu/getting-started/additional-resources/distributing-tasks-with-slurm-and-gnu-parallel/
https://rcc.uchicago.edu/docs/running-jobs/srun-parallel/index.html

Of course syntax may need to be tweaked, options added, etc, but it will be useful for users to have a somewhat familiar syntax when transitioning to Flux. (With the added benefit of being to use parallel-like facilities across an entire instance)

In fact there may even be cases where Flux could be a compelling alternative to standalone GNU parallel.

grondo commented:
I pushed a few more enhancements here that might be useful in the short term:

 * A `--shuffle` option which randomizes the submission of jobs in `bulksubmit`
 * Support for some special input delimiters inspired by GNU parallel:
    - `:::+`: "link" the input with other inputs instead of combining. That is, instead of generating all combinations of this input with other inputs, cycle through the inputs assigning one to each final input, e.g.
        ```
        $ flux mini bulksubmit --dry-run {},{} ::: A B C D E :::+ 1 2
        flux-mini: submit A,1
        flux-mini: submit B,2
        flux-mini: submit C,1
        flux-mini: submit D,2
        flux-mini: submit E,1
        ```
    - `::::`: Read the next set of inputs from a file. This can be used to combine `stdin` with other inputs by specifying a filename of `-`.
    - `::::+`: Read from file, but link input list instead of combining it
   
 * Support replacement in some missed options,  e.g. `--time-limit` and `--env`. With `--env` `bulksubmit` can simulate its own `--cc` option for example:
    ```
    $ flux mini bulksubmit --label-io -q --watch --env=FLUX_JOB_CC={seq} printenv FLUX_JOB_CC ::: {1..4}
    ƒ4Nkpq1ouZ: 0: 0
    ƒ4NkprVoBu: 0: 1
    ƒ4NkprVoBv: 0: 2
    ƒ4NkprVoBw: 0: 3
   ````


Here's a somewhat useful example of generating 100 `sleep 1` jobs ranging from 1-4 tasks each with a random urgency between 1-31:

```
$ flux mini bulksubmit --dry-run --urgency={2} -n {1} sleep 1 ::: {1..100} :::+ {1..4} :::+ $(shuf -i 1-31 -n 100) | head -10
flux-mini: submitting a total of 100 jobs
flux-mini: submit -n1 --urgency=3 sleep 1
flux-mini: submit -n2 --urgency=5 sleep 1
flux-mini: submit -n3 --urgency=13 sleep 1
flux-mini: submit -n4 --urgency=10 sleep 1
flux-mini: submit -n1 --urgency=28 sleep 1
flux-mini: submit -n2 --urgency=25 sleep 1
flux-mini: submit -n3 --urgency=29 sleep 1
flux-mini: submit -n4 --urgency=11 sleep 1
flux-mini: submit -n1 --urgency=8 sleep 1
$ flux mini bulksubmit -q --jps --progress --urgency={2} -n {1} sleep 1 ::: {1..100} :::+ {1..4} :::+ $(shuf -i 1-31 -n 100)
Submitting 100 jobs: │███████████████████████████████████████│100.0% 319.8 job/s
```

grondo commented:
I requested at least a cursory review from @SteVwonder to catch anything obvious.

However, it would be good to get this in soon, since I could use these interfaces to good effect when developing tests for #3464.

garlick commented:
I like the new options, and well done on the documentation.  I agree this should go in soon.

grondo commented:
Ok, I think I'll just set MWP on this one now, so the interfaces can be used in ongoing work.

Any issues with the underlying code can always be fixed in future PRs.

SteVwonder commented:
Thanks @grondo and sorry for the delay. This is an awesome addition! Love the GNU parallel-style syntax!
